### ISSUE TITLE
Gradle sync failed: already finished

### ISSUE THREAD

**Issue by Ali-Razmjoo** [guardianproject/orbot#117]
Hello,

I wanted to compile this project with android studio on windows, but I've got an error `Gradle sync failed: already finished`

![photo5963163243109264186](https://user-images.githubusercontent.com/7676267/34880986-6e4805d2-f7c7-11e7-9dc6-3077b75af2fb.jpg)

I've tried several ways which I found in google, but it won't solve the problem! (clean project, remove caches and ...)

Could you please give me some guides?

Regards.

**Comment by n8fr8** [comment]
It looks like you imported it incorrectly. There should be multiple projects, one for the app, one for orbotservice, one for jsocksandroid.

That said, we don't do development on Windows, only on Linux and Mac. 

Can you try building with gradlew from the command line first?

**Comment by Ali-Razmjoo** [comment]
Hello,

Sorry for the delay in answering. I imported it wrong. BTW I've got some new errors. would you please give me some guide on this one too?

![photo5965467561783045116](https://user-images.githubusercontent.com/7676267/34916227-e5d2a4da-f949-11e7-8897-144bdf4ff5c9.jpg)
![photo5965467561783045117](https://user-images.githubusercontent.com/7676267/34916228-e5fe58d2-f949-11e7-97f6-810629fbf773.jpg)


Regards.

**Comment by dovanvu1792** [comment]
I have same error "Gradle sync failed: already finished" I use macOSX :(( . @n8fr8 Can you help me to fix it?

**Comment by bitmold** [comment]
@Ali-Razmjoo it seems like you haven't built the native code based off of libtun2socks being missing 

You first need to run `git submodule update --init --recursive` in the root directory of the repository and then from there navigate to `/orbotservice/src/main` and run `ndk-build`.

**Comment by Ali-Razmjoo** [comment]
Hello @bitmold,

I've already done all of the guides in the [BUILD](https://github.com/n8fr8/orbot/blob/master/BUILD) file.

Regards.

**Comment by goapunk** [comment]
@Ali-Razmjoo 
>I've already done all of the guides in the BUILD file.

Check if `orbotservice/src/main/libs/armeabi/libtun2socks.so` exists. If not please post the output of `ndk-build`

@dovanvu1792 Did you try to re-import it?

**Comment by Ali-Razmjoo** [comment]
Hello, @goapunk,

the file was exist, and here is the output.

```
D:\android sample4\orbot\orbotservice\src\main>NDK-BUILD
Android NDK: WARNING: APP_PLATFORM android-16 is higher than android:minSdkVersion 1 in ./AndroidManifest.xml. NDK binaries will *not* be comptible with devices older than android-16. See https://android.googlesource.com/platform/ndk/+/master/docs/user/common_problems.md for more information.
The syntax of the command is incorrect.
process_begin: CreateProcess(NULL, zip assets/armeabi/pdnsd.mp3 libs/armeabi/pdnsd, ...) failed.
The syntax of the command is incorrect.
process_begin: CreateProcess(NULL, zip assets/armeabi/pdnsd.mp3 libs/armeabi/pdnsd, ...) failed.
[armeabi] Compile thumb  : pdnsd <= cache.c
jni/pdnsd/src/cache.c: In function 'read_disk_cache':
jni/pdnsd/src/cache.c:1386:2: warning: implicit declaration of function 'stpcpy' [-Wimplicit-function-declaration]
  stpcpy(stpcpy(path,global.cache_dir),"/pdnsd.cache");
  ^
jni/pdnsd/src/cache.c:1386:2: warning: incompatible implicit declaration of built-in function 'stpcpy'
jni/pdnsd/src/cache.c: In function 'write_disk_cache':
jni/pdnsd/src/cache.c:1609:2: warning: incompatible implicit declaration of built-in function 'stpcpy'
  stpcpy(stpcpy(path,global.cache_dir),"/pdnsd.cache");
  ^
[armeabi] Compile thumb  : pdnsd <= conf-parser.c
jni/pdnsd/src/conf-parser.c: In function 'confparse':
jni/pdnsd/src/conf-parser.c:685:10: warning: variable 'ignore_cd' set but not used [-Wunused-but-set-variable]
      int ignore_cd;
          ^
[armeabi] Compile thumb  : pdnsd <= conff.c
jni/pdnsd/src/conff.c: In function 'read_config_file':
jni/pdnsd/src/conff.c:157:1: warning: label 'close_file' defined but not used [-Wunused-label]
 close_file:
 ^
[armeabi] Compile thumb  : pdnsd <= consts.c
[armeabi] Compile thumb  : pdnsd <= debug.c
[armeabi] Compile thumb  : pdnsd <= dns.c
[armeabi] Compile thumb  : pdnsd <= dns_answer.c
[armeabi] Compile thumb  : pdnsd <= dns_query.c
jni/pdnsd/src/dns_query.c: In function 'p_exec_query':
jni/pdnsd/src/dns_query.c:1403:22: warning: variable 'qc' set but not used [-Wunused-but-set-variable]
    unsigned short qt,qc;
                      ^
[armeabi] Compile thumb  : pdnsd <= error.c
[armeabi] Compile thumb  : pdnsd <= hash.c
[armeabi] Compile thumb  : pdnsd <= helpers.c
[armeabi] Compile thumb  : pdnsd <= icmp.c
[armeabi] Compile thumb  : pdnsd <= list.c
jni/pdnsd/src/list.c: In function 'dlist_grow':
jni/pdnsd/src/list.c:113:3: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
   *((size_t *)&a->data[a->last])=a->lastsz;
   ^
jni/pdnsd/src/list.c:128:3: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
   *((size_t *)&a->data[sz])=0;
   ^
[armeabi] Compile thumb  : pdnsd <= main.c
jni/pdnsd/src/main.c: In function 'main':
jni/pdnsd/src/main.c:572:4: warning: implicit declaration of function 'stpcpy' [-Wimplicit-function-declaration]
    stpcpy(stpcpy(dbgpath,global.cache_dir),"/pdnsd.debug");
    ^
jni/pdnsd/src/main.c:572:4: warning: incompatible implicit declaration of built-in function 'stpcpy'
[armeabi] Compile thumb  : pdnsd <= netdev.c
jni/pdnsd/src/netdev.c: In function 'dev_up':
jni/pdnsd/src/netdev.c:142:4: warning: implicit declaration of function 'stpcpy' [-Wimplicit-function-declaration]
    stpcpy(stpcpy(stpcpy(path,"/var/run/"),ifname),".pid");
    ^
jni/pdnsd/src/netdev.c:142:4: warning: incompatible implicit declaration of built-in function 'stpcpy'
jni/pdnsd/src/netdev.c:155:4: warning: incompatible implicit declaration of built-in function 'stpcpy'
    stpcpy(stpcpy(path,"/var/lock/LCK.."),devname);
    ^
[armeabi] Compile thumb  : pdnsd <= rr_types.c
[armeabi] Compile thumb  : pdnsd <= servers.c
[armeabi] Compile thumb  : pdnsd <= status.c
jni/pdnsd/src/status.c: In function 'init_stat_sock':
jni/pdnsd/src/status.c:779:2: warning: implicit declaration of function 'stpcpy' [-Wimplicit-function-declaration]
  stpcpy(stpcpy(sa->sun_path,global.cache_dir),"/pdnsd.status");
  ^
jni/pdnsd/src/status.c:779:2: warning: incompatible implicit declaration of built-in function 'stpcpy'
[armeabi] Compile thumb  : pdnsd <= thread.c
[armeabi] Executable     : pdnsd
[armeabi] Install        : pdnsd => libs/armeabi/pdnsd
[armeabi] Compile thumb  : tun2socks <= BLog_syslog.c
[armeabi] Compile thumb  : tun2socks <= BReactor_badvpn.c
[armeabi] Compile thumb  : tun2socks <= BSignal.c
[armeabi] Compile thumb  : tun2socks <= BConnection_unix.c
[armeabi] Compile thumb  : tun2socks <= BTime.c
[armeabi] Compile thumb  : tun2socks <= BUnixSignal.c
[armeabi] Compile thumb  : tun2socks <= BNetwork.c
[armeabi] Compile thumb  : tun2socks <= StreamRecvInterface.c
[armeabi] Compile thumb  : tun2socks <= PacketRecvInterface.c
[armeabi] Compile thumb  : tun2socks <= PacketPassInterface.c
[armeabi] Compile thumb  : tun2socks <= StreamPassInterface.c
[armeabi] Compile thumb  : tun2socks <= SinglePacketBuffer.c
[armeabi] Compile thumb  : tun2socks <= BufferWriter.c
[armeabi] Compile thumb  : tun2socks <= PacketBuffer.c
[armeabi] Compile thumb  : tun2socks <= PacketStreamSender.c
[armeabi] Compile thumb  : tun2socks <= PacketPassConnector.c
[armeabi] Compile thumb  : tun2socks <= PacketProtoFlow.c
[armeabi] Compile thumb  : tun2socks <= PacketPassFairQueue.c
[armeabi] Compile thumb  : tun2socks <= PacketProtoEncoder.c
[armeabi] Compile thumb  : tun2socks <= PacketProtoDecoder.c
[armeabi] Compile thumb  : tun2socks <= BSocksClient.c
[armeabi] Compile thumb  : tun2socks <= BTap.c
[armeabi] Compile thumb  : tun2socks <= timers.c
[armeabi] Compile thumb  : tun2socks <= udp.c
[armeabi] Compile thumb  : tun2socks <= memp.c
[armeabi] Compile thumb  : tun2socks <= init.c
[armeabi] Compile thumb  : tun2socks <= pbuf.c
[armeabi] Compile thumb  : tun2socks <= tcp.c
[armeabi] Compile thumb  : tun2socks <= tcp_out.c
[armeabi] Compile thumb  : tun2socks <= netif.c
jni/../../../../external/badvpn/lwip/src/core/netif.c: In function 'netif_create_ip6_linklocal_address':
jni/../../../../external/badvpn/lwip/src/core/netif.c:895:68: warning: iteration 6u invokes undefined behavior [-Waggressive-loop-optimizations]
       netif->ip6_addr[0].addr[addr_index] |= ((u32_t)(netif->hwaddr[netif->hwaddr_len - i - 1])) << (8 * (i & 0x03));
                                                                    ^
jni/../../../../external/badvpn/lwip/src/core/netif.c:891:5: note: containing loop
     for (i = 0; i < 8; i++) {
     ^
[armeabi] Compile thumb  : tun2socks <= def.c
[armeabi] Compile thumb  : tun2socks <= mem.c
[armeabi] Compile thumb  : tun2socks <= tcp_in.c
[armeabi] Compile thumb  : tun2socks <= stats.c
[armeabi] Compile thumb  : tun2socks <= inet_chksum.c
[armeabi] Compile thumb  : tun2socks <= icmp.c
[armeabi] Compile thumb  : tun2socks <= igmp.c
[armeabi] Compile thumb  : tun2socks <= ip4_addr.c
[armeabi] Compile thumb  : tun2socks <= ip_frag.c
[armeabi] Compile thumb  : tun2socks <= ip4.c
[armeabi] Compile thumb  : tun2socks <= autoip.c
[armeabi] Compile thumb  : tun2socks <= ethip6.c
[armeabi] Compile thumb  : tun2socks <= inet6.c
[armeabi] Compile thumb  : tun2socks <= ip6_addr.c
[armeabi] Compile thumb  : tun2socks <= mld6.c
[armeabi] Compile thumb  : tun2socks <= dhcp6.c
[armeabi] Compile thumb  : tun2socks <= icmp6.c
[armeabi] Compile thumb  : tun2socks <= ip6.c
[armeabi] Compile thumb  : tun2socks <= ip6_frag.c
[armeabi] Compile thumb  : tun2socks <= nd6.c
[armeabi] Compile thumb  : tun2socks <= sys.c
[armeabi] Compile thumb  : tun2socks <= tun2socks.c
[armeabi] Compile thumb  : tun2socks <= DebugObject.c
[armeabi] Compile thumb  : tun2socks <= BLog.c
[armeabi] Compile thumb  : tun2socks <= BPending.c
[armeabi] Compile thumb  : tun2socks <= BDatagram_unix.c
[armeabi] Compile thumb  : tun2socks <= PacketPassInactivityMonitor.c
[armeabi] Compile thumb  : tun2socks <= SocksUdpGwClient.c
[armeabi] Compile thumb  : tun2socks <= UdpGwClient.c
[armeabi] Compile thumb  : ancillary <= fd_recv.c
[armeabi] Compile thumb  : ancillary <= fd_send.c
[armeabi] StaticLibrary  : libancillary.a
[armeabi] SharedLibrary  : libtun2socks.so
[armeabi] Install        : libtun2socks.so => libs/armeabi/libtun2socks.so
[x86] Compile        : pdnsd <= cache.c
jni/pdnsd/src/cache.c: In function 'read_disk_cache':
jni/pdnsd/src/cache.c:1386:2: warning: implicit declaration of function 'stpcpy' [-Wimplicit-function-declaration]
  stpcpy(stpcpy(path,global.cache_dir),"/pdnsd.cache");
  ^
jni/pdnsd/src/cache.c:1386:2: warning: incompatible implicit declaration of built-in function 'stpcpy'
jni/pdnsd/src/cache.c: In function 'write_disk_cache':
jni/pdnsd/src/cache.c:1609:2: warning: incompatible implicit declaration of built-in function 'stpcpy'
  stpcpy(stpcpy(path,global.cache_dir),"/pdnsd.cache");
  ^
[x86] Compile        : pdnsd <= conf-parser.c
jni/pdnsd/src/conf-parser.c: In function 'confparse':
jni/pdnsd/src/conf-parser.c:685:10: warning: variable 'ignore_cd' set but not used [-Wunused-but-set-variable]
      int ignore_cd;
          ^
[x86] Compile        : pdnsd <= conff.c
jni/pdnsd/src/conff.c: In function 'read_config_file':
jni/pdnsd/src/conff.c:157:1: warning: label 'close_file' defined but not used [-Wunused-label]
 close_file:
 ^
[x86] Compile        : pdnsd <= consts.c
[x86] Compile        : pdnsd <= debug.c
[x86] Compile        : pdnsd <= dns.c
[x86] Compile        : pdnsd <= dns_answer.c
[x86] Compile        : pdnsd <= dns_query.c
jni/pdnsd/src/dns_query.c: In function 'p_exec_query':
jni/pdnsd/src/dns_query.c:1403:22: warning: variable 'qc' set but not used [-Wunused-but-set-variable]
    unsigned short qt,qc;
                      ^
[x86] Compile        : pdnsd <= error.c
[x86] Compile        : pdnsd <= hash.c
[x86] Compile        : pdnsd <= helpers.c
[x86] Compile        : pdnsd <= icmp.c
[x86] Compile        : pdnsd <= list.c
jni/pdnsd/src/list.c: In function 'dlist_grow':
jni/pdnsd/src/list.c:113:3: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
   *((size_t *)&a->data[a->last])=a->lastsz;
   ^
jni/pdnsd/src/list.c:128:3: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
   *((size_t *)&a->data[sz])=0;
   ^
[x86] Compile        : pdnsd <= main.c
jni/pdnsd/src/main.c: In function 'main':
jni/pdnsd/src/main.c:572:4: warning: implicit declaration of function 'stpcpy' [-Wimplicit-function-declaration]
    stpcpy(stpcpy(dbgpath,global.cache_dir),"/pdnsd.debug");
    ^
jni/pdnsd/src/main.c:572:4: warning: incompatible implicit declaration of built-in function 'stpcpy'
[x86] Compile        : pdnsd <= netdev.c
jni/pdnsd/src/netdev.c: In function 'dev_up':
jni/pdnsd/src/netdev.c:142:4: warning: implicit declaration of function 'stpcpy' [-Wimplicit-function-declaration]
    stpcpy(stpcpy(stpcpy(path,"/var/run/"),ifname),".pid");
    ^
jni/pdnsd/src/netdev.c:142:4: warning: incompatible implicit declaration of built-in function 'stpcpy'
jni/pdnsd/src/netdev.c:155:4: warning: incompatible implicit declaration of built-in function 'stpcpy'
    stpcpy(stpcpy(path,"/var/lock/LCK.."),devname);
    ^
[x86] Compile        : pdnsd <= rr_types.c
[x86] Compile        : pdnsd <= servers.c
[x86] Compile        : pdnsd <= status.c
jni/pdnsd/src/status.c: In function 'init_stat_sock':
jni/pdnsd/src/status.c:779:2: warning: implicit declaration of function 'stpcpy' [-Wimplicit-function-declaration]
  stpcpy(stpcpy(sa->sun_path,global.cache_dir),"/pdnsd.status");
  ^
jni/pdnsd/src/status.c:779:2: warning: incompatible implicit declaration of built-in function 'stpcpy'
[x86] Compile        : pdnsd <= thread.c
[x86] Executable     : pdnsd
[x86] Install        : pdnsd => libs/x86/pdnsd
[x86] Compile        : tun2socks <= BLog_syslog.c
[x86] Compile        : tun2socks <= BReactor_badvpn.c
[x86] Compile        : tun2socks <= BSignal.c
[x86] Compile        : tun2socks <= BConnection_unix.c
[x86] Compile        : tun2socks <= BTime.c
[x86] Compile        : tun2socks <= BUnixSignal.c
[x86] Compile        : tun2socks <= BNetwork.c
[x86] Compile        : tun2socks <= StreamRecvInterface.c
[x86] Compile        : tun2socks <= PacketRecvInterface.c
[x86] Compile        : tun2socks <= PacketPassInterface.c
[x86] Compile        : tun2socks <= StreamPassInterface.c
[x86] Compile        : tun2socks <= SinglePacketBuffer.c
[x86] Compile        : tun2socks <= BufferWriter.c
[x86] Compile        : tun2socks <= PacketBuffer.c
[x86] Compile        : tun2socks <= PacketStreamSender.c
[x86] Compile        : tun2socks <= PacketPassConnector.c
[x86] Compile        : tun2socks <= PacketProtoFlow.c
[x86] Compile        : tun2socks <= PacketPassFairQueue.c
[x86] Compile        : tun2socks <= PacketProtoEncoder.c
[x86] Compile        : tun2socks <= PacketProtoDecoder.c
[x86] Compile        : tun2socks <= BSocksClient.c
[x86] Compile        : tun2socks <= BTap.c
[x86] Compile        : tun2socks <= timers.c
[x86] Compile        : tun2socks <= udp.c
[x86] Compile        : tun2socks <= memp.c
[x86] Compile        : tun2socks <= init.c
[x86] Compile        : tun2socks <= pbuf.c
[x86] Compile        : tun2socks <= tcp.c
[x86] Compile        : tun2socks <= tcp_out.c
[x86] Compile        : tun2socks <= netif.c
jni/../../../../external/badvpn/lwip/src/core/netif.c: In function 'netif_create_ip6_linklocal_address':
jni/../../../../external/badvpn/lwip/src/core/netif.c:895:68: warning: iteration 6u invokes undefined behavior [-Waggressive-loop-optimizations]
       netif->ip6_addr[0].addr[addr_index] |= ((u32_t)(netif->hwaddr[netif->hwaddr_len - i - 1])) << (8 * (i & 0x03));
                                                                    ^
jni/../../../../external/badvpn/lwip/src/core/netif.c:891:5: note: containing loop
     for (i = 0; i < 8; i++) {
     ^
[x86] Compile        : tun2socks <= def.c
[x86] Compile        : tun2socks <= mem.c
[x86] Compile        : tun2socks <= tcp_in.c
[x86] Compile        : tun2socks <= stats.c
[x86] Compile        : tun2socks <= inet_chksum.c
[x86] Compile        : tun2socks <= icmp.c
[x86] Compile        : tun2socks <= igmp.c
[x86] Compile        : tun2socks <= ip4_addr.c
[x86] Compile        : tun2socks <= ip_frag.c
[x86] Compile        : tun2socks <= ip4.c
[x86] Compile        : tun2socks <= autoip.c
[x86] Compile        : tun2socks <= ethip6.c
[x86] Compile        : tun2socks <= inet6.c
[x86] Compile        : tun2socks <= ip6_addr.c
[x86] Compile        : tun2socks <= mld6.c
[x86] Compile        : tun2socks <= dhcp6.c
[x86] Compile        : tun2socks <= icmp6.c
[x86] Compile        : tun2socks <= ip6.c
[x86] Compile        : tun2socks <= ip6_frag.c
[x86] Compile        : tun2socks <= nd6.c
[x86] Compile        : tun2socks <= sys.c
[x86] Compile        : tun2socks <= tun2socks.c
[x86] Compile        : tun2socks <= DebugObject.c
[x86] Compile        : tun2socks <= BLog.c
[x86] Compile        : tun2socks <= BPending.c
[x86] Compile        : tun2socks <= BDatagram_unix.c
[x86] Compile        : tun2socks <= PacketPassInactivityMonitor.c
[x86] Compile        : tun2socks <= SocksUdpGwClient.c
[x86] Compile        : tun2socks <= UdpGwClient.c
[x86] Compile        : ancillary <= fd_recv.c
[x86] Compile        : ancillary <= fd_send.c
[x86] StaticLibrary  : libancillary.a
[x86] SharedLibrary  : libtun2socks.so
[x86] Install        : libtun2socks.so => libs/x86/libtun2socks.so
```
Regards.

**Comment by Ali-Razmjoo** [comment]
Hello, 

@dovanvu1792, Yes I was tested in on Linux and it had no problem, so last night was searching for the problem on windows and I've found it. there are two problems.

* the first one, windows is insensitive with file/folder names, there is a file named `BUILD` in the home directory ([BUILD](https://github.com/n8fr8/orbot/blob/master/BUILD)) that means no folder could be created in the windows with the same name. even `build` or `Build`. so I renamed `BUILD` to `BUILD.md`.

* second, in the path `orbot/orbotservice/src/main/ there is a file named `jniLibs` which is not visible and accessible on the Windows and it's a symlink to Libs folder. so I removed the file and create a directory with its name and copy the Libs directory content into it.

let me know if you want to fix it on windows, I can send a PR or more details.

Thank you for your helping.

Regards.
